### PR TITLE
[dashboard-oop] Launch dashboard as exe

### PR DESCRIPTION
**Pull request by mitchdenny**
Draft, do not merge. My intention of this PR is to bring together a few different moving parts. Currently there are some things hardcoded that shouldn't be and some rough edges that we don't want on the feature branch.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/1717)

### CODE REVIEW

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs [hunk 1]
```diff
@@ -85,7 +91,36 @@ public async Task RunApplicationAsync(CancellationToken cancellationToken = defa
...
+        var dashboardExecutableSpec = new ExecutableSpec();
+        dashboardExecutableSpec.ExecutionType = ExecutionType.Process;
+   
+        // HACK: Just doing this in this branch until the workload mechanism is in place
+        //       and I can emulate it.
+        dashboardExecutableSpec.ExecutablePath = "C:\\Code\\dotnet\\aspire\\artifacts\\bin\\Aspire.Dashboard\\Debug\\net8.0\\Aspire.Dashboard.exe";
+        dashboardExecutableSpec.WorkingDirectory = "C:\\Code\\dotnet\\aspire\\src\\Aspire.Dashboard";
+
+        var environment = new List<EnvVar>();
+        environment.Add(new EnvVar()
```
Comment (davidfowl) [inline src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:107]: You also need to set the URL of the dashboard itself.

Comment (mitchdenny) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:107]: Yep. We'll grab these values from the dashboard service host.

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs [hunk 2]
```diff
@@ -93,21 +97,75 @@ public async Task RunApplicationAsync(CancellationToken cancellationToken = defa
         }
     }
 
+    private bool TryGetMetadataPath(string key, [NotNullWhen(true)]out string? path)
+    {
+        var entryAssembly = Assembly.GetEntryAssembly()!; // Null suppression because it can't return null in this context.
```
Comment (davidfowl) [inline src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:102]: You should use the assembly in `DistributedApplicationOptions`. Move the processing where we process all of the other attributes.

File: src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs [hunk 3]
```diff
@@ -23,7 +23,7 @@ namespace Aspire.Hosting.Dashboard;
 internal sealed class DashboardServiceHost : IHostedService
 {
     private const string DashboardServiceUrlVariableName = "DOTNET_DASHBOARD_GRPC_ENDPOINT_URL";
-    private const string DashboardServiceUrlDefaultValue = "http://localhost:18999";
+    private const string DashboardServiceUrlDefaultValue = "https://localhost:18999";
```
Comment (davidfowl) [inline src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs:26]: Why is this https?

File: samples/eShopLite/AppHost/WorkloadAttributes.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Reflection;
+
+// In our released product this attribute is added automatically by the SDK, here we reference
+// the path relative to the AppHost working directory and code inside the ApplicationExecutor
+// resolves it to the full path.
+[assembly:AssemblyMetadata("aspiredashboardpath", "..\\..\\..\\artifacts\\bin\\Aspire.Dashboard\\Debug\\net8.0\\Aspire.Dashboard.dll")]
```
Comment (davidfowl) [inline samples/eShopLite/AppHost/WorkloadAttributes.cs:9]: This should be absolute not relative, is this just for testing purposes?

File: samples/eShopLite/AppHost/Program.cs [hunk 5]
```diff
@@ -32,6 +32,10 @@
 builder.AddProject<Projects.CatalogDb>("catalogdbapp")
        .WithReference(catalogDb);
 
-builder.AddProject<Projects.Aspire_Dashboard>("aspire-dashboard");
+// This is added to the sample project to allow for debugging of the dashboard code,
+// not required under normal circumstances.
+builder.AddProject<Projects.Aspire_Dashboard>("aspire-dashboard")
+    .WithEnvironment("DOTNET_DASHBOARD_OTLP_ENDPOINT_URL", "http://localhost:16031")
+    .WithEnvironment("DOTNET_DASHBOARD_GRPC_ENDPOINT_URL", "https://localhost:18999");
```
Comment (davidfowl) [inline samples/eShopLite/AppHost/Program.cs:39]: Why is this needed?

File: src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs (hunk 3 above)
Comment (mitchdenny) [reply src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs:26]: gRPC HTTP/2

File: samples/eShopLite/AppHost/WorkloadAttributes.cs (hunk 4 above)
Comment (mitchdenny) [reply samples/eShopLite/AppHost/WorkloadAttributes.cs:9]: Just for testing :)

Comment (mitchdenny) [reply samples/eShopLite/AppHost/WorkloadAttributes.cs:9]: The workload targets will emit a fully qualified path.

File: src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs (hunk 3 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs:26]: Should be HTTP still hardcoded to H2 as the protocol (we do this today).

File: src/Aspire.Dashboard/Model/DashboardClient.cs [hunk 6]
```diff
@@ -24,7 +24,7 @@ namespace Aspire.Dashboard.Model;
 internal sealed class DashboardClient : IDashboardClient
 {
     private const string DashboardServiceUrlVariableName = "DOTNET_DASHBOARD_GRPC_ENDPOINT_URL";
-    private const string DashboardServiceUrlDefaultValue = "http://localhost:18999";
+    private const string DashboardServiceUrlDefaultValue = "https://localhost:18999";
```
Comment (davidfowl) [inline src/Aspire.Dashboard/Model/DashboardClient.cs:27]: This should be http too

File: samples/eShopLite/AppHost/Program.cs [hunk 7]
```diff
@@ -32,6 +32,6 @@
 builder.AddProject<Projects.CatalogDb>("catalogdbapp")
        .WithReference(catalogDb);
 
-builder.AddProject<Projects.Aspire_Dashboard>("dashboard");
+builder.AddProject<Projects.Aspire_Dashboard>("aspire-dashboard");
```
Comment (JamesNK) [inline samples/eShopLite/AppHost/Program.cs:35]: This is only for the local playground, right? Could you add a comment to that effect. Make it clearer to anyone reading our code (working on aspire or using it) that adding the dashboard isn't typical.

File: src/Aspire.Dashboard/Model/DashboardClient.cs (hunk 6 above)
Comment (davidfowl) [reply src/Aspire.Dashboard/Model/DashboardClient.cs:27]: Actually, the client shouldn't have a fallback @drewnoakes

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs [hunk 8]
```diff
@@ -73,6 +78,11 @@ public async Task RunApplicationAsync(CancellationToken cancellationToken = defa
         AspireEventSource.Instance.DcpModelCreationStart();
         try
         {
+            if (!_model.Resources.Any(r => r.Name == "aspire-dashboard"))
```
Comment (JamesNK) [inline src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:81]: If this name is special, then make a constant. A better solution could be to have internal metadata to avoid accidental conflict using name.

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs [hunk 9]
```diff
@@ -85,7 +95,90 @@ public async Task RunApplicationAsync(CancellationToken cancellationToken = defa
...
+    {
+        if (!TryGetMetadataPath(DashboardPathMetadataKey, out var dashboardPath))
+        {
+            _logger.LogWarning("Dashboard is not being started because path was not present or invalid.");
+            return;
+        }
+
+        var dashboardWorkingDirectory = Path.GetDirectoryName(dashboardPath);
+
+        var dashboardExecutableSpec = new ExecutableSpec();
```
Comment (JamesNK) [inline src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:143]: Why did you decide to build the ExecutableSpec directly vs creating an executable resource, then converting it to a spec?

I could imagine it is useful to have some logic like logging and OTEL env vars automatically set - which is done at the resource layer - without lots of duplication.

File: samples/eShopLite/AppHost/Program.cs (hunk 7 above)
Comment (mitchdenny) [reply samples/eShopLite/AppHost/Program.cs]: Once we've finished with the bulk of the review I'll go through all the playgrounds and add a reference to the dashboard as well.

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs [hunk 10]
```diff
@@ -85,7 +96,57 @@ public async Task RunApplicationAsync(CancellationToken cancellationToken = defa
...
+        environment.Add(new EnvVar()
+        {
+            Name = "ASPNETCORE_ENVIRONMENT",
+            Value = "Development"
+        });
+
+        dashboardExecutableSpec.Env = environment;
+
+        var dashboardExecutable = new Executable(dashboardExecutableSpec);
+        dashboardExecutable.Metadata.Name = "aspire-dashboard";
```
Comment (JamesNK) [inline src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:147]: What happens if someone has a resource called `aspire-dashboard`?

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs (hunk 8 above)
Comment (mitchdenny) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:82]: You mean an annotation? The problem is that the playground projects would need to be able to see that annotation.

File: src/Aspire.Hosting/Dcp/DcpOptions.cs [hunk 11]
```diff
@@ -12,6 +12,7 @@ internal sealed class DcpOptions
     private const string DcpCliPathMetadataKey = "DcpCliPath";
     private const string DcpExtensionsPathMetadataKey = "DcpExtensionsPath";
     private const string DcpBinPathMetadataKey = "DcpBinPath";
+    private const string DashboardPathMetadataKey = "aspiredashboardpath";
```
Comment (JamesNK) [inline src/Aspire.Hosting/Dcp/DcpOptions.cs:15]: Different casing from other metadata keys triggers my OCD. Is there a reason?

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs (hunk 9 above)
Comment (mitchdenny) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:111]: Because we don't want the dashboard showing up as a resource. There is presently a bug in the dashboard where it is showing the aspire-dashboard executable because it is querying the DCP resource list instead of using the app model resources.

File: src/Aspire.Hosting/Dcp/DcpOptions.cs (hunk 11 above)
Comment (mitchdenny) [reply src/Aspire.Hosting/Dcp/DcpOptions.cs:15]: Provide this feedback on this PR :)

https://github.com/dotnet/aspire/pull/1705

I'm just lining up with what is over on that PR.

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs (hunk 10 above)
Comment (mitchdenny) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:147]: We won't launch the dashboard from the workload.

File: src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs (hunk 3 above)
Comment (JamesNK) [reply src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs:26]: > gRPC HTTP/2

Only HTTP/2 happens on this port. Pre-negotiated HTTP/2 (both the client and server are configured to only support HTTP/2) is supported without TLS.

Comment (mitchdenny) [reply src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs:26]: Fixed.

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs (hunk 9 above)
Comment (JamesNK) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:111]: > showing up as a resource

Is this a UI thing? What about adding support for an internal flag that says a resource shouldn't be displayed in the dashboard UI.

Comment (JamesNK) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:113]: Maybe what this PR is doing now is the right thing, but I'd like to understand why it needs special treatment.

File: samples/eShopLite/AppHost/Program.cs [hunk 12]
```diff
@@ -32,6 +32,11 @@
 builder.AddProject<Projects.CatalogDb>("catalogdbapp")
        .WithReference(catalogDb);
 
-builder.AddProject<Projects.Aspire_Dashboard>("dashboard");
+// This project is only added in playground projects to support development/debugging
+// of the dashboard. It is not required in end developer code. Comment out this code
+// to test end developer dashboard launch experience. Refer to WorkloadAttributes.cs
+// for the path to the dashboard binary (defaults to a relative path to the artifacts
+// directory).
+//builder.AddProject<Projects.Aspire_Dashboard>("aspire-dashboard");
```
Comment (davidfowl) [inline samples/eShopLite/AppHost/Program.cs:40]:
```suggestion
builder.AddProject<Projects.Aspire_Dashboard>("aspire-dashboard");
```

We want this behavior locally. Especially in this project.

File: src/Aspire.Dashboard/appsettings.Development.json [hunk 13]
```diff
@@ -4,5 +4,6 @@
       "Default": "Information",
       "Microsoft.AspNetCore": "Warning"
     }
-  }
+  },
+  "DetailedErrors": true
```
Comment (davidfowl) [inline src/Aspire.Dashboard/appsettings.Development.json:8]: ?

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs (hunk 9 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:113]: We just need to filter it out on the resource server side of things.

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs [hunk 14]
```diff
@@ -85,7 +96,59 @@ public async Task RunApplicationAsync(CancellationToken cancellationToken = defa
...
+        });
+        environment.Add(new EnvVar()
+        {
+            Name = "DOTNET_DASHBOARD_OTLP_ENDPOINT_URL",
+            Value = otlpEndpointUrl
+        });
+        environment.Add(new EnvVar()
+        {
+            Name = "ASPNETCORE_ENVIRONMENT",
+            Value = "Development"
```
Comment (davidfowl) [inline src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:143]: This should be read from the environment as well no?

File: src/Aspire.Hosting/DistributedApplicationBuilder.cs [hunk 15]
```diff
@@ -63,7 +63,8 @@ public DistributedApplicationBuilder(DistributedApplicationOptions options)
         _innerBuilder.Services.AddHostedService<DcpHostService>();
 
         // Dashboard
-        _innerBuilder.Services.AddHostedService<DashboardServiceHost>();
+        _innerBuilder.Services.AddSingleton<DashboardServiceHost>();
+        _innerBuilder.Services.AddHostedService<DashboardServiceHost>(sp => sp.GetRequiredService<DashboardServiceHost>());
```
Comment (davidfowl) [inline src/Aspire.Hosting/DistributedApplicationBuilder.cs:67]: Why this change?

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs (hunk 9 above)
Comment (mitchdenny) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:113]: The logic is that the dashboard is not a resource as far as the app model is concerned. It is more like `dcp.exe`. It doesn't get referenced by other resources, in fact it needs to be spun up before other resources to avoid missing telemetry (side note, this PR does not yet contain a check to make sure that the dashboard has started).

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs [hunk 16]
```diff
@@ -73,6 +79,11 @@ public async Task RunApplicationAsync(CancellationToken cancellationToken = defa
         AspireEventSource.Instance.DcpModelCreationStart();
         try
         {
+            if (!_model.Resources.Any(r => r.Name == DashboardReservedResourceName))
+            {
+                await StartDashboardAsync(cancellationToken).ConfigureAwait(false);
+            }
```
Comment (davidfowl) [inline src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:85]: This logic should always start the dashboard first. Even if it's a project resource (when debugging locally).

File: samples/eShopLite/AppHost/Program.cs (hunk 12 above)
Comment (mitchdenny) [reply samples/eShopLite/AppHost/Program.cs:40]: Yeah ... I'll change this back once I've finished testing things out :)

File: src/Aspire.Dashboard/appsettings.Development.json (hunk 13 above)
Comment (mitchdenny) [reply src/Aspire.Dashboard/appsettings.Development.json:8]: This will disappear, this change was made on main recently and its useful for debugging issues.

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs (hunk 9 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:113]: The resource server currently queries DCP so this doesn't do anything AFAIK. Doesn't it still show up? If you still need to filter it out, what does this buy you?

Comment (mitchdenny) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:113]: It still shows up because its a bug. Dashboard shouldn't be querying DCP directly, it should be filtered through the lens of the app model. Why is it important for the dashboard to be an Aspire resource?

Comment (mitchdenny) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:113]: If we want to go down the path of making this a resource we might want to add another lifeycle callback method (something like `AfterStartAsync(...)`) where we can inject resourecs after the `DashboardServiceHost` is started but before we start allocating endpoints. Then once we've got the dashboard in the resource list - when we go to start them in DCP we would need to order them (or pick out the ones with aspire-dashboard naming specifically) and start them before the others so that the OTLP endpoint is available.

I already have to add the logic to check that the dashboard is alive before allowing execution to continue.

Comment (davidfowl) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:113]: > It still shows up because its a bug. Dashboard shouldn't be querying DCP directly, it should be filtered through the lens of the app model. Why is it important for the dashboard to be an Aspire resource?

I'm not sure if it is, but that'll make it easy to debug etc 😄. I'm OK making it a DCP resource directly, its not like the app needs to interact with it. That said, we are not changing the resource server logic to do the right thing and use the app model yet. Not until we get this end to end working as well as preview2 with all of the refactoring.

File: src/Aspire.Dashboard/appsettings.Development.json (hunk 13 above)
Comment (drewnoakes) [reply src/Aspire.Dashboard/appsettings.Development.json:8]: @davidfowl it seems like this is a helpful setting to have in the development appsettings. It causes the browser to show more detailed exception messages when the Blazor circuit fails. Any concern here?

File: src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs [hunk 17]
```diff
@@ -68,36 +75,64 @@ internal sealed class DashboardServiceHost : IHostedService
...
+            return false;
+        }
+
+        if (!Uri.TryCreate(value, UriKind.Absolute, out var candidateUri))
+        {
+            uri = null;
+            return false;
+        }
+
+        if (candidateUri.Scheme.ToLowerInvariant() != "https")
```
Comment (drewnoakes) [inline src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs:120]: Avoid a potential allocation:

```suggestion
        if (!StringComparer.InvariantCultureIgnoreCase.Equals(candidateUri.Scheme, "https"))
```

File: src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs [hunk 18]
```diff
@@ -68,36 +75,64 @@ internal sealed class DashboardServiceHost : IHostedService
...
+        {
+            uri = null;
+            return false;
+        }
+
+        if (candidateUri.Host.ToLowerInvariant() != "localhost")
```
Comment (drewnoakes) [inline src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs:126]: Avoid a potential allocation:

```suggestion
        if (!StringComparer.InvariantCultureIgnoreCase.Equals(candidateUri.Host, "localhost"))
```

Or maybe just:

```suggestion
        if (!candidateUri.IsLoopback)
```

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs [hunk 19]
```diff
@@ -73,6 +79,11 @@ public async Task RunApplicationAsync(CancellationToken cancellationToken = defa
         AspireEventSource.Instance.DcpModelCreationStart();
         try
         {
+            if (!_model.Resources.Any(r => r.Name == DashboardReservedResourceName))
```
Comment (drewnoakes) [inline src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:82]:
```suggestion
            if (!_model.Resources.Any(r => StringComparers.ResourceNames.Equals(r.Name, DashboardReservedResourceName)))
```

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs [hunk 20]
```diff
@@ -85,7 +96,60 @@ public async Task RunApplicationAsync(CancellationToken cancellationToken = defa
         {
             AspireEventSource.Instance.DcpModelCreationStop();
         }
+    }
+
+    private const string DashboardReservedResourceName = "aspire-dashboard";
```
Comment (drewnoakes) [inline src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:101]: Consider creating:

```c#
internal static class KnownResourceNames
{
    public const string DashboardReservedResourceName = "aspire-dashboard";
}
```

We have some `Known*` classes already that follow this pattern and allow reuse across projects.

File: src/Aspire.Dashboard/appsettings.Development.json (hunk 13 above)
Comment (davidfowl) [reply src/Aspire.Dashboard/appsettings.Development.json:8]: Seems fine as long as we don't run the dashboard development outside of the repo.

File: src/Aspire.Hosting/DistributedApplicationBuilder.cs (hunk 15 above)
Comment (mitchdenny) [reply src/Aspire.Hosting/DistributedApplicationBuilder.cs:67]: I use it to grab the dashboard service URI (gRPC endpoint) from the DashboardServiceHost. My long standing gripe about AddHostedService is that it also doesn't inject it as a singleton service :) This is my workaround!

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs (hunk 9 above)
Comment (mitchdenny) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:113]: This PR now includes logic to filter out aspire-dashboard* resources. Still some tidy up to do on that but its working.

File: src/Aspire.Hosting/Dashboard/DcpDataSource.cs [hunk 21]
```diff
@@ -86,6 +91,24 @@ await foreach (var (eventType, resource) in _kubernetesService.WatchAsync<T>(can
...
     }
 
+    private static bool IsFilteredResource<T>(T resource) where T: CustomResource
+    {
+        // We filter out any resources that start with aspire-dashboard (there are services as well as executables).
+        if (resource.Metadata.Name.StartsWith("aspire-dashboard", StringComparisons.ResourceName))
+        {
+            if (Environment.GetEnvironmentVariable("DOTNET_ASPIRE_SHOW_DASHBOARD_RESOURCES") is { } showDashboardResourcesValue)
+            {
+                return bool.TryParse(showDashboardResourcesValue, out var parsedShowDashboardResourcesValue) && !parsedShowDashboardResourcesValue;
```
Comment (drewnoakes) [inline src/Aspire.Hosting/Dashboard/DcpDataSource.cs:101]: Some VS env vars use `1` and `0`. Is there precedent on only accepting valid bool strings?

File: src/Aspire.Hosting/Dcp/DcpOptions.cs [hunk 22]
```diff
@@ -31,6 +32,8 @@ internal sealed class DcpOptions
     /// </example>
     public string? ExtensionsPath { get; set; }
 
+    public string? DashboardPath { get; set; }
```
Comment (drewnoakes) [inline src/Aspire.Hosting/Dcp/DcpOptions.cs:35]: Would be great to see API doc on this member, similar to that of the prior two properties, with an example value.

**Comment by drewnoakes** [comment]
I pushed a fix for a unit test.

The [most recent run failed](https://dev.azure.com/dnceng-public/public/_build/results?buildId=536856&view=results) because the tests in  `Aspire.Microsoft.EntityFrameworkCore.SqlServer.Tests.dll [net8.0|x64]` never completed. Kicking off another build in case it's transient.

**Comment by drewnoakes** [comment]
Yep, it was transient.

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs (hunk 20 above)
Comment (drewnoakes) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:103]: Addressed in #1775

File: src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs
[the remark this answers is not in the document]
Comment (drewnoakes) [reply src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs:1]: Mitch and I merged these two PRs.

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs (hunk 16 above)
Comment (drewnoakes) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:87]: The local debug case is addressed in #1775

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs (hunk 8 above)
Comment (drewnoakes) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs]: Made constant in #1775

File: src/Aspire.Hosting/Dcp/DcpOptions.cs (hunk 22 above)
Comment (drewnoakes) [reply src/Aspire.Hosting/Dcp/DcpOptions.cs:35]: Addressed in #1777
